# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

In [1]:
from scipy.stats import binom, geom, poisson, norm, expon
import matplotlib.pyplot as plt
import numpy as np

In [5]:
n = 460
p = 0.97

binom_dist = binom(n, p)

binom_dist.cdf(450)

0.8844772466215431

88.45% of the time, the company will have enough seats to accommodate all passengers who show up (despite overbooking by 10 extra tickets).

**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

In [7]:
p = 0.3

geom_dist = geom(p)

1 - geom_dist.cdf(2)

0.49

geom_dist.cdf(k) = P(X≤k) = the probability that success occurs at most on the kth attempt.

There is a 49% chance that the representative will have to try at least 3 times before successfully resolving the issue.

**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

In [10]:
mu = 550

poisson_dist = poisson (mu)
poisson_dist.pmf(500)

0.0017099017831047655

In [11]:
mu = 500 # average
k_max = 550 # maximum capacity

# probability that the server is not overwhelmed
p_not_overwhelmed = poisson.cdf(k_max, mu)

# probability that the server is overwhelmed
p_overwhelmed = 1 - p_not_overwhelmed

print("Probability that the server is overwhelmed:", p_overwhelmed)

Probability that the server is overwhelmed: 0.01289822084039205


What is the probability of being overwhelmed at some point during a day? (consider 24hours)

In [9]:
mu = 12000 # 300x24 for one day

poisson_dist = poisson (mu)
poisson_dist.pmf(13200)

2.044346719971809e-28

**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

In [14]:
# rate of events per unit of time 
# 1 customer per 10 min ---> 0.1 customer arrives per 10 min 

lambda_value = 0.1

expon_dist = expon(scale = 1/lambda_value)

expon_dist.cdf(5)

0.3934693402873666

There is about a 39% chance that the next customer will arrive within the next 5 minutes.

If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

In [15]:
lambda_rate = 0.1  # per minute
scale = 1 / lambda_rate  # scale parameter for expon
t = 15

# create exponential distribution
expon_dist = expon(scale=scale)

# probability no customer arrives for 15 minutes
p_break = 1 - expon_dist.cdf(t)
print("Probability an employee can take a break:", p_break)


Probability an employee can take a break: 0.2231301601484298


**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

In [17]:
mean = 150 
std = 10

norm_dist = norm(loc = mean, scale = std)

In [19]:
p = norm_dist.cdf(160) - norm_dist.cdf(140)
print("Probability that weight is between 140 and 160:", p)

Probability that weight is between 140 and 160: 0.6826894921370859


68.3% of birds of this species weigh between 140g and 160g.

**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

In [21]:
mean = 50

p = expon(scale=mean).cdf(30)
p


0.4511883639059736